lib/ansible/modules/storage/glusterfs/gluster_volume.py

```diff
@@ -465,22 +511,28 @@ def main():
             new_bricks = []
             removed_bricks = []
             all_bricks = []
+            bricks_in_volume = volumes[volume_name]['bricks']
+
             for node in cluster:
                 for brick_path in brick_paths:
                     brick = '%s:%s' % (node, brick_path)
                     all_bricks.append(brick)
-                    if brick not in volumes[volume_name]['bricks']:
+                    if brick not in bricks_in_volume:
                         new_bricks.append(brick)
 
-            # this module does not yet remove bricks, but we check those anyways
-            for brick in volumes[volume_name]['bricks']:
-                if brick not in all_bricks:
-                    removed_bricks.append(brick)
+            if not new_bricks and len(all_bricks) < bricks_in_volume:
+                for brick in bricks_in_volume:
+                    if brick not in all_bricks:
+                        removed_bricks.append(brick)
 
             if new_bricks:
                 add_bricks(volume_name, new_bricks, stripes, replicas, force)
                 changed = True
 
+            if removed_bricks:
```

Comment:
Shouldn't there be a validation on the number of bricks being removed? @sac @devyanikota ?
For instance, if replica count =3, then 3 bricks need to be removed, if removing a sub-volume. Otherwise, replica count needs to be reduced..etc. Is remove-brick supported for plain distribute - shouldn't data migration be triggered?

Reply:
@sabose we do not validate the bricks since validation is done by gluster. We'd be duplicating the efforts.
Plain distribute is supported, data migration is triggered when we remove the brick. However, commit is tried 12 times with 10 second gap. If the data migration is not complete by then, we give up and it is upto the user to commit. This is because depending on the data migrated it might sometimes take days together to complete.

Reply:
Isn't there a difference in how remove-brick is called for replica vs distribute? For distribute volume types - it's remove-brick start followed by remove-brick commit. However if reducing replica count, there's no need for data migration, and hence it's remove-brick force. Does the module account for it? (I'm not sure of the logic of disperse volume types - please check that as well)

Reply:
I tested the remove-brick operation on various gluster volume types(replicated, distributed, arbiter, disperse).
I was able to successfully remove bricks from the volumes as long as the number of bricks to be removed were of the correct configuration(a multiple of the brick config) of the respective volume type.
@sac @sabose ^^
Resources used:
[Gluster_Admin_guide](https://access.redhat.com/documentation/en-US/Red_Hat_Storage/2.0/html/Administration_Guide/sect-User_Guide-Managing_Volumes-Shrinking.html)
[Gluster docs](https://docs.gluster.org/en/v3/Administrator%20Guide/Managing%20Volumes/#shrinking-volumes)

Reply:
ok! @devyanikota so there's no need for the module to append force based on volume types? It's handled  by gluster?